### PR TITLE
NEXT-10842 - Allow creating orders via the REST API

### LINKED ISSUE #120: NEXT-10842 - Allow creating entities containing `StateMachineStateField`s via the REST API
**Description**
It is currently not possible to create an instance of any entity that contains a field of type `StateMachineStateField` (e.g. `Order::stateId`) using the admin REST API. Attempts to do so result in an error:
```json
{
  "errors": [
    {
      "code": "FRAMEWORK__WRITE_CONSTRAINT_VIOLATION",
      "detail": "This field is write-protected. (Got: \"user\" scope and \"system\" is required)",
      "meta": {
        "parameters": [
          "user",
          "system"
        ]
      },
      "source": {
        "pointer": "/0/stateIdstateId"
      },
      "status": "400",
      "template": "This field is write-protected. (Got: \"%s\" scope and \"%s\" is required)"
    }
  ]
}
```

It should be possible though to create any entity via the REST API, no matter what fields they have.

**Possible Solution**
There are a few solutions for this:
1. Define a default state for each entity which is applied by the REST API while not allowing to set a set.
2. Disable the write validation for REST API writes.
3. **Preferred solution** based on 1.: Use the `StateMachineRegistry` to validate and apply the state passed to the REST API after the entity itself was created to ensure the correct transition events are fired etc.

**Additional context**
There is already an issue for this regarding the order creation: https://issues.shopware.com/issues/NEXT-7800

### PR DESCRIPTION
**Pull request by windaishi**
### 1. What does this change do, exactly?
This PR removed the default write-protection of the `StateMachineState` field. The check, whether the change of the field is allowed is now done in a newly created `StateMachineStateFieldSerializer`. There it is possible to distinguish whether the entity is being created or edited. Then a corresponding error can be thrown.

### 2. Describe each step to reproduce the issue or behaviour.
See Issue https://issues.shopware.com/issues/NEXT-10842 or https://issues.shopware.com/issues/NEXT-7800

### 3. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/master/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### Jira Issues affected:

* https://issues.shopware.com/issues/NEXT-10842
* https://issues.shopware.com/issues/NEXT-7800

Closes #120

### CODE REVIEW

File: src/Core/Framework/DataAbstractionLayer/FieldSerializer/StateMachineStateFieldSerializer.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<?php
+// Copyright (c) Pickware GmbH. All rights reserved.
```
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/FieldSerializer/StateMachineStateFieldSerializer.php:2]: Copyright header :)

Comment (windaishi) [reply src/Core/Framework/DataAbstractionLayer/FieldSerializer/StateMachineStateFieldSerializer.php:2]: Ahh, right, thx :D Anyway, this PR is still in draft, so not ready for review

**Comment by windaishi** [comment]
Some test are failing, I will have a look at this.

**Comment by windaishi** [comment]
@shyim Should be good to go now.

**Review by shyim — APPROVED** [review]

**Pull request merged**